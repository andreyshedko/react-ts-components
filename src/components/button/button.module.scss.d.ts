export const button: string;
export const warning: string;
export const info: string;
export const danger: string;
export const withIcon: string;
export const badge: string;
