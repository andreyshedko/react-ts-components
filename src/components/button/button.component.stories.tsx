import React from 'react';
import { actions } from '@storybook/addon-actions';
import { faCoffee } from '@fortawesome/free-solid-svg-icons';

import { ButtonComponent } from './button.component';

export default {
  component: ButtonComponent,
  title: 'Button Component',
};

const eventsFromNames = actions('onClick', 'onKeyDown', 'onFocus', 'onBlur');

export const severityInfo = () =>
  <ButtonComponent
    severity="info" text={"Test"}
    {...eventsFromNames} />;

export const severityWarning = () =>
  <ButtonComponent
    severity="warning"
    text={"Test"}
    {...eventsFromNames} />

export const severityDanger = () =>
  <ButtonComponent
    severity="danger"
    text={"Test"}
    {...eventsFromNames} />

export const withIcon = () =>
  <ButtonComponent
    icon={faCoffee}
    severity="info"
    text={"Test"}
    {...eventsFromNames} />;

export const disabledButton = () =>
  <ButtonComponent
    icon={faCoffee}
    severity="info"
    disabled={true}
    text={"Test"}
    {...eventsFromNames} />;