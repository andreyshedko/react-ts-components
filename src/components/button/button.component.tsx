import * as React from "react";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { IconDefinition } from '@fortawesome/free-solid-svg-icons';

import styles from './button.module.css';

export interface ButtonComponentProps {
  text: string;
  severity: string;
  disabled?: boolean;
  icon?: IconDefinition;
  onClick: (e: MouseEvent) => void;
  onKeyDown: (e: KeyboardEvent) => void;
}

export const ButtonComponent: React.FC<ButtonComponentProps> = (props: ButtonComponentProps) => (
  <button
    className={[styles.button, styles[props.severity]].join(' ')}
    onClick={(e: React.MouseEvent<HTMLButtonElement>): void => props.onClick(e as any)}
    onKeyDown={(e: React.KeyboardEvent<HTMLElement>): void => props.onKeyDown(e as any)}
    // onBlur={}
    // onFocus={}
    disabled={props.disabled ?? false}>
    {props.icon &&
      <FontAwesomeIcon icon={props.icon} />
    }
    {props.text}
  </button>
);