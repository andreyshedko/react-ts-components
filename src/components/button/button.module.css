@import "../../variables.css";

.button {
    color: var(--gray-3);
    border: 1px solid var(--gray-1);
    border-radius: calc(var(--sm) / 6);
    padding: calc(var(--sm) / 2);
    cursor: pointer;
    background-color: #fff;
    display: flex;
    flex-flow: row nowrap;
    align-items: center;
    position: relative;
}

button svg {
    margin-right: .5rem;
}

.button:hover {
    background-color: #f4f4f4;
}

.button:disabled {
    color: var(--gray-1);
    border: 1px solid var(--gray-1);
}

.button:disabled:hover {
    cursor: not-allowed;
}

.warning {
    color: var(--warning);
    border: 1px solid var(--warning);
}

.button.warning:focus {
   outline: none;
}

.button.info:focus {
    outline: none;
 }

 .button.danger:focus {
    outline: none;
 }

.button:focus {
    background-color: #f0f0fb;
}

.danger {
    color: var(--danger);
    border: 1px solid var(--danger);
}

.info {
    color: var(--info);
    border: 1px solid var(--info);
}

.with-icon svg {
    margin-right: 0.5rem;
}

.button .badge {
    position: absolute;
    top: -10px;
    right: -10px;
    padding: 2px 5px;
    border-color: #fff;
    border-radius: 50%;
    background-color: var(--info);
    color: white;
}

.button > .badge + .info {
    background-color: var(--info);
}

.button > .badge + .warning {
    background-color: var(--warning);
}

.button .badge .danger {
    background-color: var(--danger);
}
