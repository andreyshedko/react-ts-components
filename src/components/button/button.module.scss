@import "../../variables.scss";

.button {
    color: $gray-3;
    border: 1px solid $gray-1;
    border-radius: $sm / 6;
    padding: $sm / 2;
    cursor: pointer;
    background-color: $default;
    display: flex;
    flex-flow: row nowrap;
    align-items: center;
    position: relative;

    &:hover {
        background-color: #f4f4f4;
    }

    &:disabled {
        color: $gray-1;
        border: 1px solid $gray-1;
    }

    &:disabled:hover {
        cursor: not-allowed;
    }

    svg {
        margin-right: .5rem;
    }

    & .warning {
        color: $warning;
        border: 1px solid $warning;
        &:focus {
            outline: none;
        }
    }

    & .info {
        color: $info;
        border: 1px solid $info;
        :focus {
            outline: none;
        }
    }

    & .danger {
        color: $danger;
        border: 1px solid $danger;
        &:focus {
            outline: none;
        }
    }

    & .with-icon svg {
        margin-right: 0.5rem;
    }

    & .badge {
        position: absolute;
        top: -10px;
        right: -10px;
        padding: 2px 5px;
        border-color: $default;
        border-radius: 50%;
        background-color: $info;
        color: $default;

        > .info {
            background-color: $info;
        }
    
        > .warning {
            background-color: $warning;
        }
    
        > .danger {
            background-color: $danger;
        }
    }
}