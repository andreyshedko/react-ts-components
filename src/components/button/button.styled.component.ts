
import styled from "styled-components";

type NotificationProps = {
    readonly severity: 'info' | 'warning' | 'danger',
    readonly disabled?: boolean
}

const severity = new Map([
    ['info', '#007bff'],
    ['warning', '#ff9f1c'],
    ['danger', '#e63946']]
)

export const Button = styled.button<NotificationProps>`
    color: #fff;
    border: none;
    border-radius: calc(1rem / 6);
    padding: calc(1rem / 2);
    cursor: pointer;
    background-color: ${props => props.disabled ? '#bfbfbf' : severity.get(props.severity)};
    display: flex;
    flex-flow: row nowrap;
    align-items: center;
    position: relative;
    &:hover {
        opacity: 0.8;
    }

    &:disabled {
        color: #7b7b7b;
        border: 1px solid #7b7b7b;
    }

    &:disabled:hover {
        cursor: not-allowed;
    }

    svg {
        margin-right: .5rem;
    }
`;

export const Notification = styled.span<NotificationProps>`
    position: absolute;
    top: -10px;
    right: -10px;
    padding: 2px 5px;
    border: 1px solid #fff;
    border-radius: 50%;
    color: #fff;
    background-color: ${props => severity.get(props.severity)};
`;