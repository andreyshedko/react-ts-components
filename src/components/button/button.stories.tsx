import React from 'react';
import { action } from '@storybook/addon-actions';
import { ButtonComponent } from './button.component';

export default {
  component: ButtonComponent,
  title: 'Button'
};

export const severityInfo = () => <ButtonComponent severity="info" text={"Test"} onClick={action('clicked')} onKeyDown={action('key down')} />;

export const severityWarning = () =>  <ButtonComponent severity="warning" text={"Test"} onClick={action('clicked')} onKeyDown={action('key down')} />