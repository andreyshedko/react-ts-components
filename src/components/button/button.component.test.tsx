import React from "react";

import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { ButtonComponent } from "./button.component";

configure({ adapter: new Adapter() });

const mockFn = jest.fn();

describe("Button Component", () => {
  test("should display info severity state", async () => {
    const component = renderer.create(
      <ButtonComponent
        severity={"info"}
        text={"test"}
        onClick={mockFn}
        onKeyDown={mockFn}
        onFocus={mockFn}
        onBlur={mockFn} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("should display warning severity state", async () => {
    const component = renderer.create(
      <ButtonComponent
        severity={"warning"}
        text={"test"}
        onClick={mockFn}
        onKeyDown={mockFn}
        onFocus={mockFn}
        onBlur={mockFn} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("should display danger severity state", async () => {
    const component = renderer.create(
      <ButtonComponent
        severity={"danger"}
        text={"test"}
        onClick={mockFn}
        onKeyDown={mockFn}
        onFocus={mockFn}
        onBlur={mockFn} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("fire click event", async () => {
    const wrapper = shallow(<ButtonComponent
      severity={"danger"}
      text={"test"}
      onClick={mockFn}
      onKeyDown={mockFn}
      onFocus={mockFn}
      onBlur={mockFn} />);
    wrapper.simulate('click');
    expect(mockFn).toHaveBeenCalled();
  });

  test("fire keydown event", async () => {
    const wrapper = shallow(<ButtonComponent
      severity={"danger"}
      text={"test"}
      onClick={() => {}}
      onKeyDown={mockFn}
      onFocus={mockFn}
      onBlur={mockFn} />);
    wrapper.simulate('keydown');
    expect(mockFn).toHaveBeenCalled();
  });

}); 