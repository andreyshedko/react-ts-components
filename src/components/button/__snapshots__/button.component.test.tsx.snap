// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should display danger severity state 1`] = `
<button
  className="button danger"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
>
  test
</button>
`;

exports[`Button Component should display info severity state 1`] = `
<button
  className="button info"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
>
  test
</button>
`;

exports[`Button Component should display warning severity state 1`] = `
<button
  className="button warning"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
>
  test
</button>
`;
